collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java

```diff
@@ -42,261 +35,127 @@
-
-    /**
-     * If this index is non-null, we will render an intermediary "picker" view
-     * showing the instances of the given repeat group.
-     */
-    private FormIndex repeatGroupPickerIndex;
-    private static final String REPEAT_GROUP_PICKER_INDEX_KEY = "REPEAT_GROUP_PICKER_INDEX_KEY";
-
-    /**
-     * The index of the question or the field list the FormController was set to when the hierarchy
-     * was accessed. Used to jump the user back to where they were if applicable.
-     */
-    private FormIndex startIndex;
-
-    /**
-     * The index of the question that is being displayed in the hierarchy. On first launch, it is
-     * the same as {@link #startIndex}. It can then become the index of a repeat instance.
-     */
-    private FormIndex currentIndex;
-
-    /**
-     * The index of the screen that is being displayed in the hierarchy
-     * (either the root of the form or a repeat group).
-     */
-    private FormIndex screenIndex;
-
-    /**
-     * The toolbar menu.
-     */
-    private Menu optionsMenu;
-
-    protected Button jumpBeginningButton;
-    protected Button jumpEndButton;
-    protected RecyclerView recyclerView;
+public class FormHierarchyFragment extends Fragment {
 
+    private final boolean viewOnly;
+    private final ViewModelProvider.Factory viewModelFactory;
+    private final MenuHost menuHost;
+    private FormHiearchyMenuProvider menuProvider;
     private FormEntryViewModel formEntryViewModel;
+    private FormHierarchyViewModel formHierarchyViewModel;
 
-    @Inject
-    Scheduler scheduler;
-
-    @Inject
-    FormSessionRepository formSessionRepository;
-
-    @Inject
-    MediaUtils mediaUtils;
+    public FormHierarchyFragment(boolean viewOnly, ViewModelProvider.Factory viewModelFactory, MenuHost menuHost) {
+        super(R.layout.form_hierarchy_layout);
+        this.viewOnly = viewOnly;
 
-    @Inject
-    Analytics analytics;
-
-    @Inject
-    AudioRecorder audioRecorder;
-
-    @Inject
-    ProjectsDataService projectsDataService;
-
-    @Inject
-    EntitiesRepositoryProvider entitiesRepositoryProvider;
-
-    @Inject
-    PermissionsChecker permissionsChecker;
-
-    @Inject
-    LocationClient fusedLocationClient;
-
-    @Inject
-    SettingsProvider settingsProvider;
+        this.viewModelFactory = viewModelFactory;
+        this.menuHost = menuHost;
+    }
 
-    @Inject
-    PermissionsProvider permissionsProvider;
+    @Override
+    public void onAttach(@NonNull Context context) {
+        super.onAttach(context);
 
-    @Inject
-    public AutoSendSettingsProvider autoSendSettingsProvider;
+        formEntryViewModel = new ViewModelProvider(requireActivity(), viewModelFactory).get(FormEntryViewModel.class);
+        formHierarchyViewModel = new ViewModelProvider(this, new ViewModelProvider.Factory() {
+            @NonNull
+            @Override
+            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
```

Comment:
We could move this to `FormEntryViewModelFactory` and just use `viewModelFactory` as above.

Reply:
`FormHierarchyViewModel` is currently private to `FormHierarchyFragment` and I reckon as we continue we'll probably remove it and move its functionality into `FormEntryViewModel`.